jcl/src/openj9.criu/share/classes/org/eclipse/openj9/criu/CRIUSupport.java

```diff
@@ -585,24 +590,27 @@ private static RestoreException throwSetEnvException(Throwable cause) {
 	 *                                       restore
 	 */
 	public synchronized void checkpointJVM() {
-		/* Add env variables restore hook. */
-		registerRestoreEnvVariables();
-
-		/* Add security provider hooks. */
-		SecurityProviders.registerResetCRIUState();
-		SecurityProviders.registerRestoreSecurityProviders();
+		if (isCRIUSupportEnabled()) {
+			if (InternalCRIUSupport.isCheckpointAllowed()) {
+				/* Add env variables restore hook. */
+				registerRestoreEnvVariables();
+				/* Add security provider hooks. */
+				SecurityProviders.registerResetCRIUState();
+				SecurityProviders.registerRestoreSecurityProviders();
 
-		if (InternalCRIUSupport.isCheckpointAllowed()) {
-			init();
-			checkpointJVMImpl(imageDir, leaveRunning, shellJob, extUnixSupport, logLevel, logFile, fileLocks, workDir,
-					tcpEstablished, autoDedup, trackMemory, unprivileged);
-		} else {
-			if (InternalCRIUSupport.isCRIUSupportEnabled()) {
-				throw new UnsupportedOperationException(
-					"Running in non-portable mode (only one checkpoint is allowed), and we have already checkpointed once"); //$NON-NLS-1$
+				try {
+					checkpointJVMImpl(imageDir, leaveRunning, shellJob, extUnixSupport, logLevel, logFile, fileLocks,
+							workDir, tcpEstablished, autoDedup, trackMemory, unprivileged);
+				} catch (UnsatisfiedLinkError ule) {
```

Comment:
is this exception still possible if we've loaded the library?

Reply:
perhaps we should throw an internalVMError instead. I think the message should say that there is something wrong with the JDK

Reply:
okay, if we load the library but fail to bind the method then the problem is in the JDK, so I think internalVMError (or something like that) would be more appropriate. The message should indicate that there is a problem with libj9criu in the JDK.

Reply:
Missed this notification, `InternalError("There is a problem with libj9criu in the JDK")` is added.